node/src/block_producer/vrf_evaluator/block_producer_vrf_evaluator_state.rs

```diff
@@ -1,72 +1,73 @@
 use std::collections::BTreeMap;
 use std::sync::Arc;
 
-use mina_p2p_messages::v2::LedgerHash;
+use mina_p2p_messages::v2::{
+    ConsensusProofOfStakeDataEpochDataNextValueVersionedValueStableV1,
+    ConsensusProofOfStakeDataEpochDataStakingValueVersionedValueStableV1, EpochSeed, LedgerHash,
+    MinaBaseEpochSeedStableV1,
+};
 use openmina_core::block::ArcBlockWithHash;
 use serde::{Deserialize, Serialize};
 
 use crate::block_producer::BlockProducerWonSlot;
 
-use super::{DelegatorTable, VrfWonSlotWithHash};
-
-// TODO(adonagy): consodilate types, make more clear
-// pub type AccountAddressAndBalance = (String, u64);
-
-// TODO(adonagy): cleanup state for production version.
-//
-// We should try to make impossible states irrepresentable.
-// Some examples:
-// - status can be: `DataPending`, yet `epoch_data.delegator_table` can
-//   be set.
-// - `current_epoch_data` and/or `next_epoch_data` might not be set, but
-//   we might be in a state which expects those to be set.
-// etc...
-//
-// Sure, enabling conditions, reducers and effects might help mitigate
-// that, but type system won't help us and it will be very easy to make
-// mistakes.
-// Also as a reader, state has no contraints so I can't deduce much from it.
-// When making changes, I'll have to review all the parts to make sure
-// hidden link between `status` and the rest of the data is being enforced
-// somewhere.
-//
-// For the very least a most of this state should be moved inside the `status`.
-// That way there is a direct link between `status` and underlying data.
-// Only thing I'd leave outside the `status` from below, is `won_slots`,
-// since we need to accumulate that.
-//
-// Even better than that would be to have 2 enums. Main one for preparing
-// data necessary for evaluation (building delegatee table), and once
-// we are done there, have a nested state for slot evaluation.
+use super::{DelegatorTable, VrfEvaluatorInput, VrfWonSlotWithHash};
+
+pub const SLOTS_PER_EPOCH: u32 = 7140;
+/// Vrf evaluator sub-state
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct BlockProducerVrfEvaluatorState {
     pub status: BlockProducerVrfEvaluatorStatus,
     pub won_slots: BTreeMap<u32, VrfWonSlotWithHash>,
-    pub current_epoch_data: Option<EpochData>,
-    pub next_epoch_data: Option<EpochData>,
-    // TODO(adonagy): move to block producer state probably
-    pub current_epoch: Option<u32>,
-    pub current_best_tip_slot: u32,
     pub latest_evaluated_slot: u32,
-    pub last_possible_evaluation_slot: u32,
     pub genesis_timestamp: redux::Timestamp,
+    last_evaluated_epoch: Option<u32>,
+    last_block_heights_in_epoch: BTreeMap<u32, u32>,
+    pending_evaluation: Option<PendingEvaluation>,
+    epoch_context: EpochContext,
```

Comment:
Still there are impossible states that could be avoided by a different design.

E.g. `pending_evaluation` could be `Some(_)` when the `status` isn't `BlockProducerVrfEvaluatorStatus::SlotEvaluationPending`.
Same for other fields.

I know it's a hustle to include those fields in enum variants and carry them across variants which does include that field, but it makes impossible states not even representable. Also in general it does lead to more restricted/simpler state machine.